language/move-vm/runtime/src/loader.rs

```diff
@@ -612,6 +603,43 @@ impl Loader {
         // module will NOT show up in `module_cache`. In the module republishing case, it means
         // that the old module is still in the `module_cache`, unless a new Loader is created,
         // which means that a new MoveVM instance needs to be created.
+        self.verify_module_verify_no_missing_dependencies(module, data_store, log_context)?;
+
+        // friendship is an upward edge in the dependencies DAG, so it has to be checked after the
+        // module is put into the bundle.
+        let friends = module.immediate_friends();
+        self.load_dependencies_verify_no_missing_dependencies(friends, data_store, log_context)?;
+        self.verify_module_cyclic_relations(module)
+
+        // NOTE: one might wonder why we don't need to worry about `module` (say M) being missing in
+        // the code cache? Obviously, if a `friend`, say module F, is being loaded and verified, and
+        // F may call into M; then M not being in the code cache will definitely lead to an error
+        // when verifying F because F depends on M.
+        //
+        // The answer is: given the current
+        // 1) *publish-one-module-at-a-time* model,
+        // 2) module compatibility checking scheme, and
+        // 3) how the code cache is maintained (insertion-only and no purging),
```

Comment:
If one of (1) - (3) changes in the future, will we notice that we now need to handle this case (e.g., via a test that fails)?

Reply:
I have added several test cases in the IR testsuite (https://github.com/diem/diem/pull/7531/commits/1d431ba9315621d935e7b5a7c2f5c730e8918c34, to be added in a follow-up PR), the ones named `ir-testsuite/tests/move/publish/republish_module_*.mvir` will be useful in checking these cases.